### DISCUSSION TITLE
Using opnetelemetry-java in context of a reactive Spring-Boot application

### DISCUSSION THREAD

**Discussion by frommeyerc** [open-telemetry/opentelemetry-java-instrumentation#4471] [Q&A]
Hi,

I'm currently evaluating the opentelemetry tracing instrumentation for use in a company setup. The current target are new Spring-Boot services. Most of them are currently being implemented using the new reactive web framework provided by Spring build on top of the Reactor library.
Using the spring-boot autoconfiguration library from a first glance works like a charm. However after looking closer I discovered that the tracing context is not properly handled. The default implementation uses a ThreadLocal which of course doesn't work in a reactor project where the thread of execution can easily switch between java threads. I tried to build a custom implementation using the Reactor context. However that turns out to be tricky. First of all without blocking I'd only get a `Mono<Context>` from a call to `Mono.defereContextual()`. This could be solved with blocking. But for attaching the Context I'd need to get hold of the Reactive Pipeline which seams to be quite hard with the given interfaces. Also this would break with the way e.g. the WebClient instrumentation currently is build.

I'm curious on what the opinion here is. Is this set of libraries going to support reactive Spring-Boot applications in the future? Is there maybe even already a solution on how to do that now? Or will these libraries only support the traditional programming model?

Cheers
Christian

**Comment by trask** [comment]
hey @frommeyerc! check out https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/reactor-3.1/library, this is standalone reactor support (which is also auto-injected by the javaagent)

**Reply by frommeyerc** [reply]
Hey @trask, I found that one. But how would I use it? There doesn't seem to be any autoconfiguration available for Spring and I wasn't able to find any documentation. I'd be happy to try it but currently I'm stuck.

**Reply by trask** [reply]
here's the artifact that you would use: https://search.maven.org/search?q=g:io.opentelemetry.instrumentation%20AND%20a:opentelemetry-reactor-3.1

at startup of your application you'll want to run `ContextPropagationOperator.create().registerOnEachOperator()`, which should set things up for you

**Reply by frommeyerc** [reply]
That doesn't seem to change anything. Every span still gets a new traceId. The context doesn't seem to get transported.

**Reply by trask** [reply]
can you post a repro that we can look at to see what's going on?

**Reply by frommeyerc** [reply]
Sure. Find my project here: https://github.com/frommeyerc/reactive-tracing-dummy

**Reply by trask** [reply]
@frommeyerc sorry for the delay, can you create an issue for this to make sure we don't lose track of it?

**Comment by frommeyerc** [comment]
Added an [issue](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4982).
